File: examples/node-agent/bot.mjs
#!/usr/bin/env node
/* eslint-disable no-console */

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function parseArgs(argv) {
  const out = { joinToken: "", pollMs: 1200, runForMs: 0, verbose: false };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (!out.joinToken && typeof a === "string" && a.startsWith("CT1|") && a.includes("|")) out.joinToken = a;
    else if (a === "--pollMs") out.pollMs = Number(argv[++i] || 0);
    else if (a === "--runForMs") out.runForMs = Number(argv[++i] || 0);
    else if (a === "--verbose") out.verbose = true;
  }
  return out;
}

function parseJoinToken(joinToken) {
  const raw = String(joinToken || "").trim();
  const parts = raw.split("|");
  if (parts.length !== 3) return null;
  const version = parts[0].trim();
  const baseUrl = parts[1].trim().replace(/\/+$/, "");
  const joinCode = parts[2].trim();
  if (version !== "CT1" || !baseUrl || !joinCode) return null;
  return { raw, baseUrl, joinCode };
}

async function apiJson(url, { method = "GET", headers = {}, body } = {}) {
  const res = await fetch(url, {
    method,
    headers: { "Content-Type": "application/json", ...headers },
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  const text = await res.text();
  let data;
  try {
    data = JSON.parse(text);
  } catch {
    data = { raw: text };
  }
  return { ok: res.ok, status: res.status, data };
}

function nearestAliveMonster(you, monsters) {
  const list = Array.isArray(monsters) ? monsters : [];
  if (!you) return null;
  let best = null;
  let bestD2 = Infinity;
  for (const m of list) {
    if (!m || m.alive === false) continue;
    const dx = Number(m.x || 0) - Number(you.x || 0);
    const dy = Number(m.y || 0) - Number(you.y || 0);
    const d2 = dx * dx + dy * dy;
    if (d2 < bestD2) {
      bestD2 = d2;
      best = m;
    }
  }
  return best ? { m: best, d2: bestD2 } : null;
}

async function main() {
  const args = parseArgs(process.argv);
  const parsed = parseJoinToken(args.joinToken);
  if (!parsed) {
    console.error('Usage: node examples/node-agent/bot.mjs "CT1|<baseUrl>|<joinCode>" [--runForMs 60000] [--pollMs 1200] [--verbose]');
    process.exit(2);
  }

  const pollMs = Math.max(500, Math.floor(Number(args.pollMs) || 1200));
  const runForMs = Math.max(0, Math.floor(Number(args.runForMs) || 0));
  const verbose = Boolean(args.verbose);

  const baseUrl = parsed.baseUrl;
  const linked = await apiJson(`${baseUrl}/api/bot/link`, { method: "POST", body: { joinToken: parsed.raw } });
  if (!linked.ok || !linked.data?.ok || !linked.data?.botToken) {
    console.error("link failed", linked.status, linked.data);
    process.exit(2);
  }
  const botToken = String(linked.data.botToken);
  const headers = { Authorization: `Bearer ${botToken}` };

  const mode = await apiJson(`${baseUrl}/api/bot/mode`, { method: "POST", headers, body: { mode: "agent" } });
  if (!mode.ok || !mode.data?.ok) {
    console.error("mode failed", mode.status, mode.data);
    process.exit(2);
  }

  console.log(`Connected. baseUrl=${baseUrl} playerId=${linked.data.playerId || ""}`);
  console.log("Loop: world → goal/cast. Ctrl+C to stop.");

  const started = Date.now();
  let lastCastAt = 0;
  let lastGoalAt = 0;
  let lastThoughtAt = 0;

  // eslint-disable-next-line no-constant-condition
  while (true) {
    if (runForMs > 0 && Date.now() - started > runForMs) break;

    const w = await apiJson(`${baseUrl}/api/bot/world`, { headers });
    if (!w.ok || !w.data?.ok) {
      if (verbose) console.log("world error", w.status, w.data);
      await sleep(pollMs);
      continue;
    }

    const snap = w.data.snapshot || {};
    const you = snap.you || null;
    const nearest = nearestAliveMonster(you, snap.monsters || []);
    const now = Date.now();

    if (nearest && you) {
      const hitRange = 140;
      if (nearest.d2 <= hitRange * hitRange) {
        if (now - lastCastAt > 850) {
          lastCastAt = now;
          const r = await apiJson(`${baseUrl}/api/bot/cast`, { method: "POST", headers, body: { spell: "signature" } });
          if (verbose) console.log("cast", r.status, r.data?.ok ? "ok" : r.data);
        }
      } else if (now - lastGoalAt > 1200) {
        lastGoalAt = now;
        const r = await apiJson(`${baseUrl}/api/bot/goal`, {
          method: "POST",
          headers,
          body: { x: Number(nearest.m.x || 0), y: Number(nearest.m.y || 0) },
        });
        if (verbose) console.log("goal", r.status, r.data?.ok ? "ok" : r.data);
      }
    }

    // Keep it lightweight: a thought bubble every ~8s (no public chat spam).
    if (now - lastThoughtAt > 8000) {
      lastThoughtAt = now;
      const r = await apiJson(`${baseUrl}/api/bot/thought`, { method: "POST", headers, body: { text: "Auto-grinding… (tiny agent)" } });
      if (verbose) console.log("thought", r.status, r.data?.ok ? "ok" : r.data);
    }

    await sleep(pollMs);
  }

  console.log("Done.");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/python-agent/bot.py
#!/usr/bin/env python3

import json
import sys
import time
import urllib.error
import urllib.request


def parse_join_token(raw: str):
    raw = (raw or "").strip()
    parts = raw.split("|")
    if len(parts) != 3:
        return None
    version, base_url, join_code = parts[0].strip(), parts[1].strip().rstrip("/"), parts[2].strip()
    if version != "CT1" or not base_url or not join_code:
        return None
    return {"raw": raw, "baseUrl": base_url, "joinCode": join_code}


def api_json(url: str, method="GET", headers=None, body=None, timeout=10):
    h = {"Content-Type": "application/json"}
    if headers:
        h.update(headers)
    data = None
    if body is not None:
        data = json.dumps(body).encode("utf-8")
    req = urllib.request.Request(url, data=data, headers=h, method=method)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            text = resp.read().decode("utf-8", errors="replace")
            try:
                payload = json.loads(text)
            except Exception:
                payload = {"raw": text}
            return resp.status, payload
    except urllib.error.HTTPError as e:
        text = e.read().decode("utf-8", errors="replace")
        try:
            payload = json.loads(text)
        except Exception:
            payload = {"raw": text}
        return e.code, payload
    except Exception as e:
        return 0, {"ok": False, "error": str(e)}


def nearest_alive_monster(you, monsters):
    if not you:
        return None
    best = None
    best_d2 = 10**18
    for m in monsters or []:
        if not m or m.get("alive") is False:
            continue
        dx = float(m.get("x") or 0) - float(you.get("x") or 0)
        dy = float(m.get("y") or 0) - float(you.get("y") or 0)
        d2 = dx * dx + dy * dy
        if d2 < best_d2:
            best_d2 = d2
            best = m
    if not best:
        return None
    return {"m": best, "d2": best_d2}


def main(argv):
    join_token = ""
    poll_ms = 1200
    run_for_sec = 0
    verbose = False

    i = 1
    while i < len(argv):
        a = argv[i]
        if not join_token and a.startswith("CT1|") and "|" in a:
            join_token = a
        elif a == "--pollMs" and i + 1 < len(argv):
            i += 1
            poll_ms = int(float(argv[i] or "0"))
        elif a == "--runForSec" and i + 1 < len(argv):
            i += 1
            run_for_sec = int(float(argv[i] or "0"))
        elif a == "--verbose":
            verbose = True
        i += 1

    parsed = parse_join_token(join_token)
    if not parsed:
        print('Usage: python3 examples/python-agent/bot.py "CT1|<baseUrl>|<joinCode>" [--runForSec 60] [--pollMs 1200] [--verbose]', file=sys.stderr)
        return 2

    base_url = parsed["baseUrl"]
    poll_ms = max(500, int(poll_ms or 1200))
    run_for_sec = max(0, int(run_for_sec or 0))

    st, linked = api_json(f"{base_url}/api/bot/link", method="POST", body={"joinToken": parsed["raw"]})
    if st < 200 or st >= 300 or not linked.get("ok") or not linked.get("botToken"):
        print("link failed", st, linked, file=sys.stderr)
        return 2

    bot_token = str(linked["botToken"])
    headers = {"Authorization": f"Bearer {bot_token}"}

    st, mode = api_json(f"{base_url}/api/bot/mode", method="POST", headers=headers, body={"mode": "agent"})
    if st < 200 or st >= 300 or not mode.get("ok"):
        print("mode failed", st, mode, file=sys.stderr)
        return 2

    print(f"Connected. baseUrl={base_url} playerId={linked.get('playerId','')}")
    print("Loop: world → goal/cast. Ctrl+C to stop.")

    started = time.time()
    last_cast_at = 0.0
    last_goal_at = 0.0
    last_thought_at = 0.0
    hit_range = 140.0

    try:
        while True:
            if run_for_sec > 0 and (time.time() - started) > run_for_sec:
                break

            st, w = api_json(f"{base_url}/api/bot/world", headers=headers)
            if st < 200 or st >= 300 or not w.get("ok"):
                if verbose:
                    print("world error", st, w, file=sys.stderr)
                time.sleep(poll_ms / 1000.0)
                continue

            snap = w.get("snapshot") or {}
            you = snap.get("you") or None
            nearest = nearest_alive_monster(you, snap.get("monsters") or [])
            now = time.time()

            if nearest and you:
                if nearest["d2"] <= hit_range * hit_range:
                    if (now - last_cast_at) > 0.85:
                        last_cast_at = now
                        st2, r = api_json(f"{base_url}/api/bot/cast", method="POST", headers=headers, body={"spell": "signature"})
                        if verbose:
                            print("cast", st2, ("ok" if r.get("ok") else r), file=sys.stderr)
                elif (now - last_goal_at) > 1.2:
                    last_goal_at = now
                    st2, r = api_json(
                        f"{base_url}/api/bot/goal",
                        method="POST",
                        headers=headers,
                        body={"x": float(nearest["m"].get("x") or 0), "y": float(nearest["m"].get("y") or 0)},
                    )
                    if verbose:
                        print("goal", st2, ("ok" if r.get("ok") else r), file=sys.stderr)

            if (now - last_thought_at) > 8.0:
                last_thought_at = now
                api_json(f"{base_url}/api/bot/thought", method="POST", headers=headers, body={"text": "Auto-grinding… (tiny agent)"})

            time.sleep(poll_ms / 1000.0)
    except KeyboardInterrupt:
        pass

    print("Done.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

